language/move-vm/runtime/src/loader.rs

```diff
@@ -664,42 +678,53 @@ impl Loader {
             .into_iter()
             .filter(|module_id| !bundle_unverified.contains(module_id))
             .collect();
-        self.load_dependencies_verify_no_missing_dependencies(cached_friends, data_store)?;
+        self.load_dependencies_verify_no_missing_dependencies(
+            cached_friends,
+            &mut in_progress,
+            data_store,
+        )?;
         self.verify_module_cyclic_relations(module, bundle_verified, bundle_unverified)
     }
 
-    fn verify_module_verify_no_missing_dependencies(
+    fn verify_module_dependencies_verify_no_missing(
         &self,
-        module: &CompiledModule,
+        id: &ModuleId,
         bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
+        in_progress: &mut BTreeMap<ModuleId, CompiledModule>,
         data_store: &mut impl DataStore,
     ) -> VMResult<()> {
-        self.verify_module(
-            module,
-            bundle_verified,
-            &mut BTreeMap::new(),
-            data_store,
-            true,
-        )
+        self.verify_module_dependencies(id, bundle_verified, in_progress, data_store, true)
     }
 
-    fn verify_module_expect_no_missing_dependencies(
+    fn verify_module_dependencies_expect_no_missing(
         &self,
-        module: &CompiledModule,
+        id: &ModuleId,
+        bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
         in_progress: &mut BTreeMap<ModuleId, CompiledModule>,
         data_store: &mut impl DataStore,
     ) -> VMResult<()> {
-        self.verify_module(module, &BTreeMap::new(), in_progress, data_store, false)
+        self.verify_module_dependencies(id, bundle_verified, in_progress, data_store, false)
     }
 
-    fn verify_module(
+    fn verify_module_dependencies(
         &self,
-        module: &CompiledModule,
+        id: &ModuleId,
         bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
         in_progress: &mut BTreeMap<ModuleId, CompiledModule>,
         data_store: &mut impl DataStore,
         verify_no_missing_modules: bool,
     ) -> VMResult<()> {
+        // Invariant: the target module (identified by `id`) must be in the `in_progress` cache.
+        let module = in_progress.get(id).ok_or_else(|| {
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                .with_message(format!(
+                    "trying to process a module that is not bytecode verified: {}::{}",
```

Comment:
Based on the description of `in_progress` below, it sounds like the problem is not actually with the intra-module verifier (which has already been run), but with the inter-module linker. If that's the case, maybe renaming the parameter to `linking_in_progress` here + adjusting the error message would make sense?

Reply:
+1 for `linking_in_progress` as it is a more explicit description of the `in_progress` cache --- modules are put there pending for linking verification.

But the error message might be OK? What I wanted to express was that
- In order for `verify_module_dependencies` to be called, the target module must be in the `linking_in_progress` cached (identified by `id`).
- This is to mark that the target module has already passed the bytecode verifier on its own, which is required by the linking checker.
- Currently we enforce that the target module must be in `linking_in_progress` in all callsites of `verify_module_dependencies` hence, the invariant violation error.

Reply:
Got it, that makes sense. I'm ok with leaving the message as-is, but your explanation here is so good that it would be great to incorporate it in the comments.

I suppose it's also possible that the module *has* been verified, but there's a bug where we forget to add it to `in_progress`. The important thing is that whoever sees this message + traces it back to the comments understands that something unexpected has happened and has an idea of how to debug it.

Reply:
> I suppose it's also possible that the module has been verified, but there's a bug where we forget to add it to in_progress.

Ah, I see it now. It is not only about raising an error but also important to show some good diagnostic info. To be future-proof, I'll add some additional explanation to the error message.

Reply:
Error message and comments updated.